### PR TITLE
Use submodules for external repositories

### PR DESCRIPTION
**Pull request by pavanbalaji**

### CODE REVIEW

**Comment by pavanbalaji** [comment]
This is ready for testing, but the testing infrastructure needs to do `git submodule init --update --recursive` before running `autogen.sh` to load all of the submodules.

FWIW, once it's in master, cloning would simply be `git clone git@github.com:pmodels/mpich --recursive`.  We will not need the `submodule init` again at that time.

**Comment by wesbland** [comment]
Would your intention be to do this with all of the dependencies? If so, you don't have ofi here.

File: .gitmodules [hunk 1]
```diff
@@ -4,3 +4,6 @@
 [submodule "src/pm/hydra/tools/topo/hwloc/hwloc"]
 	path = src/pm/hydra/tools/topo/hwloc/hwloc
 	url = https://github.com/open-mpi/hwloc
+[submodule "confdb"]
+	path = confdb
+	url = git@github.com:pmodels/confdb.git
```
Comment (wesbland) [inline .gitmodules]: I'm not sure I agree with this one. For all the others, there could conceivably be another version of UCX/hwloc/mpl that you might want to use or they could be used in other projects. Do we really expect to have another version of confdb or it be usable elsewhere? There's a lot of MPICH-isms in there.

Comment (wesbland) [inline .gitmodules]: I'm not sure I agree with this one. For all the others, there could conceivably be another version of UCX/hwloc/mpl that you might want to use or they could be used in other projects. Do we really expect to have another version of confdb or it be usable elsewhere? There's a lot of MPICH-isms in there.

**Comment by wesbland** [comment]
I'm not sure I agree with this one. For all the others, there could conceivably be another version of UCX/hwloc/mpl that you might want to use or they could be used in other projects. Do we really expect to have another version of confdb or it be usable elsewhere? There's a lot of MPICH-isms in there.

**Comment by pavanbalaji** [comment]
@wesbland

I didn't see OFI being embedded in the current code, so I didn't add that in the PR.  Did I miss it?

We need to split out `confdb` since that's used by `hydra` and `mpl` as well, not just `mpich`.

**Comment by wesbland** [comment]
That makes sense about confdb then.

As for OFI, it _can_ be embedded. All of the hooks are already there and we build it that way sometimes. I believe @raffenet 's plan was to embed it in the releases.

**Comment by pavanbalaji** [comment]
@wesbland sure; whenever `ofi` is embedded, the plan would be to use submodules for that as well (assuming everyone signs on this PR, since it's a bit disruptive for developers).

**Comment by wesbland** [comment]
@pavanbalaji - It already is embedded. You just need to drop it in `src/mpid/ch4/netmod/ofi/libfabric`.

As for whether we agree on it. I don't have a strong opinion. @raffenet had strong opinions I think, but I imagine you guys are already discussing it.

**Comment by pavanbalaji** [comment]
I don't see it, @wesbland 

https://github.com/pmodels/mpich/tree/master/src/mpid/ch4/netmod/ofi

**Comment by wesbland** [comment]
We don't put it in there unless we want it:

https://github.com/pmodels/mpich/blob/2e9615cc669f77cb4472a0c619a97fdee00d6a70/src/mpid/ch4/netmod/ofi/subconfigure.m4#L43-L44

**Comment by wesbland** [comment]
It also needs this to do the autogen:

https://github.com/pmodels/mpich/blob/2e9615cc669f77cb4472a0c619a97fdee00d6a70/autogen.sh#L167-L169

**Comment by pavanbalaji** [comment]
Is that being manually added?  That's different from `ucx`, which is being embedded by `release.pl`.  We should be consistent between `ucx` and `ofi`, and not choose different methods for the two of them.  I'll talk to @raffenet and make them consistent.

**Comment by wesbland** [comment]
The code is already there. The choice to put it in is manual. We could add something to the `release.pl`, but when we added that embedding detection (https://github.com/pmodels/mpich/commit/887e0d30f8fe8e939a509aceed02324fef7eca38), that file didn't exist.

**Comment by pavanbalaji** [comment]
@wesbland OK, I've embedded `libfabric` too.

I need to figure out how to test it on jenkins now, since it needs an extra step.

@yfguo thoughts?

**Comment by wesbland** [comment]
Yeah. The other part is that we might often want to track the latest master of all of these things, especially for development. How should this work if we want to do that? I guess we just don't do a recursive clone?

**Comment by pavanbalaji** [comment]
No, I think all submodules need to track a specific git hash.  If we don't do that, then when I checkout an older version of mpich, I want the exact git hash of `libfabric` that it was tested with and not the latest version of `libfabric`.

**Comment by wesbland** [comment]
Yes, that's true. I was more thinking about what advice we should give to people when they want the latest of everything (as many of us will want for a our day to day development).

I agree that the repository should have a particular git commit.

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,15 @@
+[submodule "src/mpid/ch4/netmod/ucx/ucx"]
+	path = src/mpid/ch4/netmod/ucx/ucx
+	url = git://github.com/openucx/ucx
+[submodule "confdb"]
+	path = confdb
+	url = git@github.com:pmodels/confdb
```
Comment (yfguo) [inline .gitmodules]: Do we really want to use SSH-based access to the submodule repos? If I am an external contributor who does not have SSH access to these repos, I will have to manually change these to HTTPS in `.gitmodules` and remember to not included those changes to my own topic commits.

**Comment by yfguo** [comment]
@pavanbalaji For Jenkins tests, it will be just adding the `git submodule update --init --recursive` to the script. This simple change would make all jobs work with the submodules.

File: .gitmodules (hunk 2 above)
Comment (pavanbalaji) [reply .gitmodules]: Fixed.

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/all
test:jenkins/ch4/all

**Comment by pavanbalaji** [comment]
Updated the PR to pull @yfguo's changes from pmodels/confdb#1

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/all
test:jenkins/ch4/all

**Comment by yfguo** [comment]
test:jenkins/ch3/all
test:jenkins/ch4/all

**Comment by yfguo** [comment]
test:jenkins/ch3/all
test:jenkins/ch4/all

**Comment by yfguo** [comment]
test:jenkins/ch3/all
test:jenkins/ch4/all

**Comment by pavanbalaji** [comment]
I think this is good to go in.  @yfguo and @raffenet ?

**Comment by yfguo** [comment]
I think this PR is good to go.

**Comment by raffenet** [comment]
This looks OK. I think there's some updates needed in the submodules themselves to avoid "dirty" reports from git status, e.g.
```	modified:   confdb (untracked content)
	modified:   src/mpid/ch4/netmod/ucx/ucx (new commits)
	modified:   src/mpl (modified content)
	modified:   src/pm/hydra (modified content)
```
but those won't be part of this PR. Thankfully, I was able to get this working in my local repo without having to nuke or re-clone it with `git submodule update --init --recursive`.

**Comment by raffenet** [comment]
Just noticed one thing missing. `maint/release.pl` needs updating to do recursive clone.

**Comment by raffenet** [comment]
After some more thought and discussion with @yfguo , I think we need to migrate things from `.gitignore` into the submodule repos as part of this PR.

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Comment by pavanbalaji** [comment]
test:jenkins/ch3/tcp

**Pull request merged**